### ISSUE TITLE
Can't define table different from 'oauth_users'

### ISSUE THREAD

**Issue by TarasHots** [bshaffer/oauth2-server-php#547]
I have existing table with users, and want oauth server to use that table. I've created custom storage class (which implements only OAuth2\Storage\UserCredentialsInterface), and passed it to server for user_credentials storage:

`$userCredentialsStorage = new MyOAuthStorage($sm);`
`$oauthService->addStorage($userCredentialsStorage, 'user_credentials');`

But server is still trying to validate POSTed username and password using oauth_users. How it can be configured? Is this possible?
I'm also interested in `storage` property in oauth config, for now I've set my custom storage, is it correct?

**Comment by thunderyup** [comment]
To use my own users table, I've extended the storage class I'm using.  In my case, it's Pdo.  You can change other things by overriding the functions in your preferred storage method.  I hope this helps.

```php
class DoLoginPdo extends OAuth2\Storage\Pdo
{
    public function __construct($connection, $config = array())
    {
        parent::__construct($connection, $config);
        $this->config['user_table'] = 'users';
    }
}
```

Then, my initialization of the Oauth2 library looks like this:

```php
        $dsn      = 'mysql:dbname=yourdatabasename;host=localhost';
        $username = 'yourdatabaseusername';
        $password = 'yourdatabasepassword';

        try{
	        // $dsn is the Data Source Name for your database, for exmaple "mysql:dbname=my_oauth2_db;host=localhost"
	        $this->oauth2storage = new DoLoginPdo(array('dsn' => $dsn, 'username' => $username, 'password' => $password));

	        // Pass a storage object or array of storage objects to the OAuth2 server class
	        $this->oauth2server = new OAuth2\Server($this->oauth2storage);

	        // Add the "User Credentials" grant type
	        $this->oauth2server->addGrantType(new OAuth2\GrantType\UserCredentials($this->oauth2storage));

        }catch(PDOException $e){
        	// DO NOT send the password to the log files.
        	echo str_replace($password, ' *** Password Removed *** ' , $e);
        	die;
        }
```

**Comment by TarasHots** [comment]
Thank you for as quick as possible answer. Seems this is only way to do this(I mean only via adding grant type). I've found solution too, but it little bit different:

`$userCredentialsStorage = new MyOAuthStorage($sm);`
                   ` $userCredentialsGrantType = new OAuth2\GrantType\UserCredentials($userCredentialsStorage);`
                    `$oauthService->addGrantType($userCredentialsGrantType);`

Maybe not so brilliant solution, but it works fine. Nevertheless, thank you. I guess documentation should be updated about this, because this functionality is really helpful, but almost undocumented.

**Comment by bshaffer** [comment]
You should definitely be able to inject your own user storage like you've
tried to do initially. This is actually the preferred way.

Are you overriding the user storage later on? I am very interested why this
would happen.

On Thursday, March 19, 2015, TarasHots <notifications@github.com> wrote:

> Closed #547 <https://github.com/bshaffer/oauth2-server-php/issues/547>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bshaffer/oauth2-server-php/issues/547#event-259314333>
> .
>


-- 
Sincerely,
Brent Shaffer
@bshaffer <http://twitter.com/bshaffer>
